spring-rabbit/src/main/java/org/springframework/amqp/rabbit/core/AsyncRabbitTemplate.java

```diff
@@ -0,0 +1,583 @@
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.springframework.amqp.rabbit.core;
+
+import java.nio.charset.Charset;
+import java.util.UUID;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentMap;
+import java.util.concurrent.ExecutionException;
+
+import org.apache.commons.logging.Log;
+import org.apache.commons.logging.LogFactory;
+
+import org.springframework.amqp.AmqpException;
+import org.springframework.amqp.core.AmqpMessageReturnedException;
+import org.springframework.amqp.core.Message;
+import org.springframework.amqp.core.MessageListener;
+import org.springframework.amqp.core.MessagePostProcessor;
+import org.springframework.amqp.core.MessageProperties;
+import org.springframework.amqp.rabbit.connection.ConnectionFactory;
+import org.springframework.amqp.rabbit.core.RabbitTemplate.ConfirmCallback;
+import org.springframework.amqp.rabbit.core.RabbitTemplate.ReturnCallback;
+import org.springframework.amqp.rabbit.listener.SimpleMessageListenerContainer;
+import org.springframework.amqp.rabbit.support.CorrelationData;
+import org.springframework.context.SmartLifecycle;
+import org.springframework.util.Assert;
+import org.springframework.util.concurrent.ListenableFuture;
+import org.springframework.util.concurrent.SettableListenableFuture;
+
+/**
+ * Provides asynchronous send and receive operations returning a {@link ListenableFuture}
+ * allowing the caller to obtain the reply later, using {@code get()} or a callback.
+ * <p>
+ * When confirms are enabled, the future has a confirm property which is itself a
+ * {@link ListenableFuture}. If the reply is received before the publisher confirm,
+ * the confirm is discarded since the reply implicitly indicates the message was
+ * published.
+ * <p>
+ * Returned (undeliverable) request messages are presented as a
+ * {@link AmqpMessageReturnedException} cause of an {@link ExecutionException}.
+ * <p>
+ * Internally, the template uses a {@link RabbitTemplate} and a
+ * {@link SimpleMessageListenerContainer} either provided or constructed internally.
+ * If an external {@link RabbitTemplate} is provided and confirms/returns are enabled,
+ * it must not previously have had callbacks registered because this object needs to
+ * be the callback.
+ *
+ * @author Gary Russell
+ * @since 1.6
+ *
+ */
+public class AsyncRabbitTemplate<C> implements SmartLifecycle, MessageListener, ReturnCallback, ConfirmCallback {
```

Comment:
I thought about doing that, but it's not really an extension of `AmqpTemplate` it wraps one.

I thought about extracting `AsyncRabbitOperations`.

I suppose we could extend `AmqpTemplate` and delegate the other methods to the `RabbitTemplate` but that won't be much use to us because in most places we use `RabbitTemplate` directly - it supports features (confirms) that are extensions to Amqp.

We could extend `RabbitTemplate` but that would introduce a number of difficulties - particularly around confirm and return handling we'd have to add a bunch more asserts, or override the methods with `UnsupportedOperationException`.

So I decided to keep it simple.

Let me know if any of these options appeals to you.